src/planner.c

```diff
@@ -401,6 +452,15 @@ timescaledb_planner(Query *parse, int cursor_opts, ParamListInfo bound_params)
 					ts_data_node_fetcher_scan_type = ts_guc_remote_data_fetcher;
 				}
 			}
+
+			if (ts_chunk_cache == NULL)
+			{
+				reset_chunk_cache = true;
+
+				ts_chunk_cache = ChunkCache_create(CurrentMemoryContext,
```

Comment:
I notice that the `CurrentMemoryContext` is quite often used to initialize the other hashes in the PostgreSQL code, but I think this practice is dangerous and places a high demand on the caller to do make sure that the memory context is correct before calling the function. I think you would be  better off picking a dedicated memory context for this cache, for example, the `CacheMemoryContext` since that is what is used when creating hypertable caches below.

Reply:
The hypertable cache lives across the queries, the chunk cache lives inside one query only. So they are going to use different context. I can use `PortalContext` here to be more specific.

Reply:
As an aside, maybe I shouldn't call this a "cache", because this is just a hash table with some data that is needed to process the query. It doesn't need invalidation, reference counting and all this stuff than comes to mind when you say "cache".

Reply:
If there is no specific need for being able to use different contexts, that would be great. Also note that the `CacheMemoryContext` is used to allocate hypertable structures, which might be suboptimal for this use-case, but should be safe (I am assuming that `CacheMemoryContext` outlives `PortalContext`).

Reply:
> As an aside, maybe I shouldn't call this a "cache", because this is just a hash table with some data that is needed to process the query. It doesn't need invalidation, reference counting and all this stuff than comes to mind when you say "cache".

That is a good point, but most of these things can be solved with a good comment. I was not sure about the lifetime of this cache, so making that clear would be an advantage. If you cache stuff, even if just for a query, the obvious question will also be about the lifetime of the data stored in the entry, so a comment about that would be great as well.

Reply:
I renamed it to `ts_per_query_baserel_info`, and there was already a comment about the scope before the definition.

Reply:
Also changed the context to `PortalContext`. Not sure if it's better, because logically this hash table should be created in the memory context of the top-level call of `timescaledb_planner`.

Reply:
That seems to be the `MessageContext`, but is there a reason to why you do not create it in that memory context then? Are you sure that the lifetime of that context is shorter than `CacheMemoryContext` (where the hypertable pointers come from).

Reply:
From comments: 
_CacheMemoryContext --- permanent storage for relcache, catcache, and
related modules.  This will never be reset or deleted, either, so it's
not truly necessary to distinguish it from TopMemoryContext_.
It should be longer lived than portal or message context.

There is another question -- do we keep the hypertable cache itself pinned for the duration of `timescaledb_planner`. We do. Looks like the lifetimes are correct.

Reply:
`MessageContext` looks more suitable indeed, I'll use it.

Reply:
Oops, the message context is null for SPI. Looks like we'll have to return to PortalContext.

Reply:
(see the segfault here https://github.com/timescale/timescaledb/runs/5631092680?check_suite_focus=true)

Reply:
Same problem for the PortalContext... I think I'll revert to using the Current one after all.

Reply:
I find it strange that is the case. Anyway, since the hypertable data is in `CacheMemoryContext` it should be longer than any other context, so should be safe.

Reply:
To be more precise, this happens for SPI calls from a background worker, probably that's why we don't have the normal contexts.